src/common/com/intellij/plugins/haxe/build/ApiInvoker.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2000-2013 JetBrains s.r.o.
+ * Copyright 2014-2016 AS3Boyan
+ * Copyright 2014-2014 Elias Ku
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.intellij.plugins.haxe.build;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+/**
+ * Created by ebishton on 1/16/16.
+ */
+public class ApiInvoker<R> {
+
+  private Object myInstance;
+  private Method myMethodToInvoke;
+  private String myDebugName;
+
+  public ApiInvoker(Object o, String methodName, Class<?>... argTypes) {
+    Class klass;
+    myDebugName = "";
+    try {
+      if (o instanceof Class) {
+        // We're looking for a static method.
+        myInstance = null;
+        klass = (Class) o;
+      } else {
+        // We have an instance method.
+        myInstance = o;
+        klass = o.getClass();
+      }
+      myDebugName = (null == klass ? "<unknown>" : klass.getName())
+                    + "." + methodName;
+      // Don't stop of null klass... Just let the exception be thrown
+      // naturally.
+      myMethodToInvoke = klass.getDeclaredMethod(methodName, argTypes);
+    } catch (NoSuchMethodException e) {
+      throw new UnsupportedApiException("Didn't find " + myDebugName + ";", e);
+    }
+  }
+
+  public R invoke(Object... args) {
+    try {
+      return (R) myMethodToInvoke.invoke(myInstance, args);
```

Comment:
Yes, it uses reflection to get around the fact that the compiler wants to see correct syntax in all branches.  This way, we can look up the function and bind it at run-time.  It gets us away from needing two source files, at the small cost of run-time linking (and possibly failure) and the potential loss of some optimizations.